src/runtime/StateSerialization/MaybeMethodBase.cs

```diff
@@ -0,0 +1,199 @@
+using System;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Linq;
+
+namespace Python.Runtime
+{
+    [Serializable]
+    internal struct MaybeMethodBase<T> : ISerializable where T: MethodBase
+    {
+        // .ToString() of the serialized object
+        const string SerializationName = "s";
+        // The ReflectedType of the object
+        const string SerializationType = "t";
+        // Fhe parameters of the MethodBase
+        const string SerializationParameters = "p";
+        const string SerializationIsCtor = "c";
+        const string SerializationMethodName = "n";
+
+        [Serializable]
+        struct ParameterHelper : IEquatable<ParameterInfo>
+        {
+            public enum TypeModifier
+            {
+                None,
+                In,
+                Out,
+                Ref
+            }
+            public readonly string Name;
+            public readonly TypeModifier Modifier;
+
+            public ParameterHelper(ParameterInfo tp)
+            {
+                Name = tp.ParameterType.AssemblyQualifiedName;
+                Modifier = TypeModifier.None;
+
+                if (tp.IsIn)
+                {
+                    Modifier = TypeModifier.In;
+                }
+                else if (tp.IsOut)
+                {
+                    Modifier = TypeModifier.Out;
+                }
+                else if (tp.ParameterType.IsByRef)
+                {
+                    Modifier = TypeModifier.Ref;
+                }
+            }
+
+            public bool Equals(ParameterInfo other)
+            {
+                return this.Equals(new ParameterHelper(other));
+            }
+        }
+        public static implicit operator MaybeMethodBase<T> (T ob) => new MaybeMethodBase<T>(ob);
+
+        string name;
+        MethodBase info;
+
+        [NonSerialized]
+        Exception deserializationException;
+
+        public string DeletedMessage 
+        {
+            get
+            {
+                return $"The .NET {typeof(T)} {name} no longer exists. Cause: " + deserializationException?.Message ;
+            }
+        }
+
+        public T Value
+        {
+            get
+            {
+                if (info == null)
+                {
+                    throw new SerializationException(DeletedMessage, innerException: deserializationException);
+                }
+                return (T)info;
+            }
+        }
```

Comment:
Couldn't you extract this as a base class? I thought you had one in previous iterations...

Reply:
They are structs, to minimize memory allocations. I could extract some of it to another struct